deltacat/compute/compactor/steps/materialize.py

```diff
@@ -164,49 +172,66 @@ def _materialize(
         mask = pa.array(mask_pylist)
         pa_table = pa_table.filter(mask)
 
-        # appending, sorting, taking, and dropping has 2-3X latency of a
-        # single filter on average, and thus provides better average
-        # performance than repeatedly filtering the table in dedupe task index
-        # order
-        dedupe_task_indices = chain.from_iterable(dedupe_task_idx_iter_tpl)
-        pa_table = sc.append_dedupe_task_idx_col(
-            pa_table,
-            dedupe_task_indices,
-        )
-        pa_sort_keys = [(sc._DEDUPE_TASK_IDX_COLUMN_NAME, "ascending")]
-        pa_table = pa_table.take(
-            pc.sort_indices(pa_table, sort_keys=pa_sort_keys),
-        )
-        pa_table = pa_table.drop(
-            [sc._DEDUPE_TASK_IDX_COLUMN_NAME]
-        )
+        if remainder_tables:
+            assert remainder_record_count < max_records_per_output_file, \
+                f"Total records in previous remainder should not exceed {max_records_per_output_file}."
+
+            # Skip materializing tables until we have reached 'max_records_per_output_file'
+            if remainder_record_count + len(pa_table) < max_records_per_output_file:
+                remainder_tables.append(pa_table)
+                remainder_record_count += len(pa_table)
+                continue
+
+            # 'max_records_per_output_file' will be exceeded with the
+            #  current file's records + previous remainder records.
+            records_to_fit = max_records_per_output_file - remainder_record_count
+            fitted_table = pa_table.slice(length=records_to_fit)
+            remainder_tables.append(fitted_table)
```

Comment:
This version looks good. However, I still feel it could be further modularized to make it easy to understand. Remember, we also have to add indices to these records as and when they are getting materialized to support primary key index building in future. In the current state, it may be difficult to make those modifications. I have proposed some changes below. Let me know if that makes sense:

```python
# This class handles splitting the tables and letting us know when it's ready to materialize
class BatchedTables(List[pa.Table]):
    def __init__(self, max_records_per_table_multiple):
        self.__init__()
        self._tables_ready = []
        self._tables_leftover = []
        self._tables_leftover_len = 0
        self._tables_ready_len = 0
    def append(self, table: pa.Table):
        # split into two in the multiples of max_records and left over and append to
        # appropriate lists created above
        if (_tables_leftover_len > max_records):
            # time to split move some of them to _ready_tables

    def get_ready_tables() -> List:
        return self._tables_ready

    def clear_ready() -> List:
        self._tables_ready.clear()

    def get_remainder_tables() -> List:
        return self._tables_leftover
  
# In your materialize code

batched_tables = BatchedTables(max_records_per_file)

for table in tables:
    batched_tables.append(table)
    if (len(batched_tables.get_ready_tables()) > 0):
        _materialize(batched_tables.get_ready_tables())
        batched_tables.clear_ready_tables()

if (batched_tables.get_ready_tables()):
    _materialize(batched_tables.get_ready_tables())
    batched_tables.clear_ready_tables()

if (batched_tables.get_remainder_tables()):
    _materialize(batched_tables.get_remainder_tables())

```

Reply:
I like the general idea. I think there could be an unintended performance hit here when doing the split to move `self._tables_leftover` over to `self._ready_tables` too frequently or across too many tables?

Reply:
I think we can mitigate that by using `double ended queue` if required as we have to maintain table ordering.

Reply:
In the class you provided, I was more so referring to how we would need to concatenate leftover tables and split them, each time we need to shift `self._tables_leftover` over to `self._ready_tables`, since we maintain two lists of tables now. What if max records is `4000`, and each table added has 3999 records?

Reply:
Doesn't the current implementation also require you to split pyarrow table everytime in the above case? As far as shifting the table goes, you will only have shift a particular table only once, so it will not affect performance. You can optimally shift if you use `double ended queues` instead of lists.

Reply:
The current impl. splits once per table, and enforce the remainder to always be less than max records for the next iteration. In this multi-list (or deque) approach we will split once per table *and* also keep track of the leftovers, triggering a shift if it exceeds max records. 

For example
```
tables = [7999, 7999, 7999, ...]

bt.append(tables[0]) # leftover = [3999], ready = [4000] (one split)
bt.append(tables[1]) # leftover = [3998], ready = [4000, 4000, 4000] (one split + one shift)
bt.append(tables[2]) # leftover = [3997], ready = [4000, 4000, 4000, 4000, 4000] (one split + one shift)
...
```

Reply:
I think the point being is that there can be other interesting ways to have this BatchedTable class, but my question is whether or not this modularization has to be P0 itself? We could have a separate issue/PR to refactor this into a class, and even have class-specific unit tests which will be easier to write vs. writing unit tests for the materialize task as a whole.

Reply:
You may think this is not p0, but without the unit tests a small change could cause data corruption. Also, you will find it hard to unit test this code vs modularized code. That way it will be easier for someone to make changes and add new features.